### DISCUSSION TITLE
Using machine learning in NLP

### DISCUSSION THREAD

**Discussion by ray-holmium** [florisboard/florisboard#2287] [General]
@patrickgold I've been thinking lots about your comment regarding implementing machine learning into the NLP in the discussion last week here [#2284](https://github.com/florisboard/florisboard/discussions/2284) and I wanted to start an open discussion on the subject.

I think this is a really exciting idea and would want to explore it further. I think there's also some great arguments to be made for avoiding altogether, but I will let others speak to that! How are others here feeling about using machine learning/ai in the suggestion engine?

One of my primary projects right now is familiarizing myself with [Nix](https://nixos.org/), which is a linux package management system designed to be declarative for reproducibility,  and I stumbled across a project called [nixified-ai](https://github.com/nixified-ai) which aims to make open source alternatives to OpenAI simple for individuals to install and run. Here is a good [video overview](https://www.youtube.com/watch?v=wJ0D7MrPNOw) of the project by its creator.

I haven't had time to do my homework to parse if this would actually be useful at all, but at first glance it seems really promising and I thought it might give you some ideas for how you could bundle a comprehensive model in with the NLP. I also found [nix-on-droid](https://github.com/t184256/nix-on-droid) so there seems to be potential.

Please let me know if you think I should open a feature request for this.

Unfortunately until I finish my declarative install of NixOS, I'm without a machine to build and test the NLP, but will hop on board there asap.

**Comment by shtrophic** [comment]
for this you probably need to implement the native AI module with Android NDK, and for that you need to come up with a model that 1. fits into memory 2. can be quantized (since the native module will probably require this) 3. convert it to the NDK format

**Comment by Waelwindows** [comment]
Hi, thanks for the interest on this project! 

> One of my primary projects right now is familiarizing myself with [Nix](https://nixos.org/), [...] I stumbled across a project called [nixified-ai](https://github.com/nixified-ai) which aims to make open source alternatives to OpenAI simple for individuals to install and run. Here is a good [video overview](https://www.youtube.com/watch?v=wJ0D7MrPNOw) of the project by its creator.

It's nice to see other people using Nix in the wild, ~~i hope you don't look at the flake we have~~. However, I don't think it's appropriate to use in this use case. For one, Nix is notorious for it's learning curve which would alienate a lot of the core user base of FlorisBoard. Second of all, it would couple FlorisBoard to _another_ 3rd party component which would complicate the whole build situation even further, unless the entire build is managable by nix, which is a topic better left for another day.  

I personally don't believe that deploying an LLM would be an appropriate solution for FlorisBoard. LLMs are quite heavy to run which would hamper their usage on lower end devices and would drain battery life. Inference latency is also an issue as user experience would degrade if the user has to wait a second or two for their suggestions to pop up. i.e. the ideal solution should display them in near real-time as the user types.

However, this doesn't mean we should abandon AI NLP solutions at all. Deploying some, more lightweight architecture (RNN? Transformer?) would be better for the project. After all, Next Word Prediction is the bread and butter of AI Language models. I have around half a dozen papers on the topic which I have yet to read. After reading them and experimenting a bit, I believe I could make a prototype suggestion system. Figuring out how to link that system to FlorisBoard would be for later.

I think it's important that @patrickgold and the community outline what their requirements are for the solution. For example:
- Should the solution learn from the user's typing habits?
- Should the solution take into account into context the words _after_, not only before?
- Should the solution be multilingual? (i.e. same architecture for many different languages)

Currently, the most promising lead for deployment seems to be to base our solution on either [TensorFlow Lite](https://www.tensorflow.org/lite) or [ONNX Runtime](https://onnxruntime.ai/). This would entail studying the feasibility of bundling the runtime + model on bundle size, which I'm positive is doable.

I currently have prior commitments but I'll see what I can do for this
